lib/improver/psychrometric_calculations/psychrometric_calculations.py

```diff
@@ -689,15 +690,169 @@ def fill_in_high_snow_falling_levels(
+        temperature profile near the surface, and use this to estimate where
+        the snow falling level would be below sea level.
+
+        The difference between the wet bulb temperature integral at the
+        threshold and the wet bulb integral at the surface is equal to the
+        integral of the wet bulb temperature between sea level and
+        the negative height corresponding to the snow falling level. As we are
+        using a simple linear fit, we can integrate this to find an expression
+        for the extrapolated snow falling level.
+
+        The form of this expression depends on whether the linear fit crosses
+        the x-axis above or below zero.
+
+        If we have our linear fit of the form:
+
+        .. math::
+            {{wet\:bulb\:temperature} = m \\times height + c}
+
+        and let I be the wet bulb temperature integral we have found above
+        sea level.
+
+        If it crosses above zero, then the limits on the integral
+        are the snow falling level and zero and we find the following
+        expression for the snow falling level:
+
+        .. math::
+            {{snow\:falling\:level} = \\frac{c \\pm \\sqrt{
+            c^2-2 m (threshold-I)}}{-m}}
+
+        If the linear fit crosses below zero the limits on our integral are
+        the snow falling level and the point where the linear fit crosses the
+        x-axis, as only possitive wet bulb temperatures count towards the
+        integral. In this case our expression for the snow falling level is:
+
+        .. math::
+            {{snow\:falling\:level} = \\frac{c \\pm \\sqrt{
+            2 m (I-threshold)}}{-m}}
+
+        Args:
+            max_wb_integral (numpy.array):
+                The wet bulb temperature integral at sea level.
+            gradient (numpy.array):
+                The gradient of the line of best fit we are using in the
+                extrapolation.
+            intercept (numpy.array):
+                The intercept of the line of best fit we are using in the
+                extrapolation.
+            snow_falling_level (numpy.array)
+                The snow falling level array with values filled in with snow
+                falling levels calculated through extrapolation.
+            sea points (numpy.array)
+                A boolean array with True where the points are sea points.
+
+        """
+
+        # Make sure we only try to extrapolate points with a valid gradient.
+        index = (gradient != 0.0) & sea_points
+        gradient = gradient[index]
+        intercept = intercept[index]
+        max_wb_int = max_wb_integral[index]
+        snow_fl = snow_falling_level[index]
+
+        # For points where -intercept/gradient is greater than zero:
+        index2 = (-intercept/gradient >= 0.0)
+        inside_sqrt = (
+            intercept[index2]**2 - 2*gradient[index2]*(
+                self.falling_level_threshold - max_wb_int[index2]))
+        snow_fl[index2] = (
```

Comment:
Anything to note about the "+" root in both cases?

Reply:
No, except it gives weird values. I haven't had chance to give it much thought.